File: dqops/src/main/java/com/dqops/rest/models/metadata/TablePartitioningModel.java
/*
 * Copyright © 2021-Present DQOps, Documati sp. z o.o. (support@dqops.com)
 *
 * This file is licensed under the Business Source License 1.1,
 * which can be found in the root directory of this repository.
 *
 * Change Date: This file will be licensed under the Apache License, Version 2.0,
 * four (4) years from its last modification date.
 */
package com.dqops.rest.models.metadata;

import com.dqops.metadata.sources.PartitionIncrementalTimeWindowSpec;
import com.dqops.metadata.sources.PhysicalTableName;
import com.dqops.metadata.sources.TableSpec;
import com.dqops.metadata.sources.TimestampColumnsSpec;
import com.dqops.utils.docs.generators.SampleStringsRegistry;
import com.dqops.utils.docs.generators.SampleValueFactory;
import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonPropertyDescription;
import com.fasterxml.jackson.databind.PropertyNamingStrategies;
import com.fasterxml.jackson.databind.annotation.JsonNaming;
import io.swagger.annotations.ApiModel;
import lombok.Data;

/**
 * Rest model that returns the configuration of table partitioning information.
 */
@Data
@JsonInclude(JsonInclude.Include.NON_NULL)
@JsonNaming(PropertyNamingStrategies.SnakeCaseStrategy.class)
@ApiModel(value = "TablePartitioningModel", description = "Table model with objects that describe the table partitioning.")
public class TablePartitioningModel {
    @JsonPropertyDescription("Connection name.")
    private String connectionName;

    @JsonPropertyDescription("Physical table details (a physical schema name and a physical table name)")
    private PhysicalTableName target;

    @JsonPropertyDescription("Column names that store the timestamps that identify the event (transaction) timestamp and the ingestion (inserted / loaded at) timestamps. Also configures the timestamp source for the date/time partitioned data quality checks (event timestamp or ingestion timestamp).")
    private TimestampColumnsSpec timestampColumns;

    @JsonPropertyDescription("Configuration of time windows for executing partition checks incrementally, configures the number of recent days to analyze for daily partitioned tables or the number of recent months for monthly partitioned data.")
    private PartitionIncrementalTimeWindowSpec incrementalTimeWindow;

    /**
     * Boolean flag that decides if the current user can update or delete this object.
     */
    @JsonPropertyDescription("Boolean flag that decides if the current user can update or delete this object.")
    private boolean canEdit;

    /**
     * Creates a table partitioning model from a table specification by cherry-picking relevant fields.
     * @param connectionName Connection name to store in the model.
     * @param tableSpec      Source table specification.
     * @param allowEdit        The calling user can edit table partitioning.
     * @return Table partitioning model.
     */
    public static TablePartitioningModel fromTableSpecification(String connectionName, TableSpec tableSpec, boolean allowEdit) {
        return new TablePartitioningModel() {{
            setConnectionName(connectionName);
            setTarget(tableSpec.getPhysicalTableName());
            setTimestampColumns(tableSpec.getTimestampColumns());
            setIncrementalTimeWindow(tableSpec.getIncrementalTimeWindow());
            setCanEdit(allowEdit);
        }};
    }

    /**
     * Updates a table specification by copying partitioning fields.
     * @param targetTableSpec Target table specification to update.
     */
    public void copyToTableSpecification(TableSpec targetTableSpec) {
        if (this.getTimestampColumns() != null) {
            targetTableSpec.setTimestampColumns(this.getTimestampColumns());
        }
        else {
            targetTableSpec.setTimestampColumns(new TimestampColumnsSpec()); // default configuration because the object is not null
        }

        if (this.getIncrementalTimeWindow() != null) {
            targetTableSpec.setIncrementalTimeWindow(this.getIncrementalTimeWindow());
        }
        else {
            targetTableSpec.setIncrementalTimeWindow(new PartitionIncrementalTimeWindowSpec()); // default configuration because the object is not null
        }
    }

    public static class TablePartitioningModelSampleFactory implements SampleValueFactory<TablePartitioningModel> {
        @Override
        public TablePartitioningModel createSample() {
            return fromTableSpecification(
                    SampleStringsRegistry.getConnectionName(),
                    new TableSpec.TableSpecSampleFactory().createSample(),
                    true);
        }
    }
}

File: distribution/python/dqops/client/models/table_partitioning_model.py
from typing import TYPE_CHECKING, Any, Dict, List, Type, TypeVar, Union

from attrs import define as _attrs_define
from attrs import field as _attrs_field

from ..types import UNSET, Unset

if TYPE_CHECKING:
    from ..models.partition_incremental_time_window_spec import (
        PartitionIncrementalTimeWindowSpec,
    )
    from ..models.physical_table_name import PhysicalTableName
    from ..models.timestamp_columns_spec import TimestampColumnsSpec


T = TypeVar("T", bound="TablePartitioningModel")


@_attrs_define
class TablePartitioningModel:
    """Table model with objects that describe the table partitioning.

    Attributes:
        connection_name (Union[Unset, str]): Connection name.
        target (Union[Unset, PhysicalTableName]):
        timestamp_columns (Union[Unset, TimestampColumnsSpec]):
        incremental_time_window (Union[Unset, PartitionIncrementalTimeWindowSpec]):
        can_edit (Union[Unset, bool]): Boolean flag that decides if the current user can update or delete this object.
    """

    connection_name: Union[Unset, str] = UNSET
    target: Union[Unset, "PhysicalTableName"] = UNSET
    timestamp_columns: Union[Unset, "TimestampColumnsSpec"] = UNSET
    incremental_time_window: Union[Unset, "PartitionIncrementalTimeWindowSpec"] = UNSET
    can_edit: Union[Unset, bool] = UNSET
    additional_properties: Dict[str, Any] = _attrs_field(init=False, factory=dict)

    def to_dict(self) -> Dict[str, Any]:
        connection_name = self.connection_name
        target: Union[Unset, Dict[str, Any]] = UNSET
        if not isinstance(self.target, Unset):
            target = self.target.to_dict()

        timestamp_columns: Union[Unset, Dict[str, Any]] = UNSET
        if not isinstance(self.timestamp_columns, Unset):
            timestamp_columns = self.timestamp_columns.to_dict()

        incremental_time_window: Union[Unset, Dict[str, Any]] = UNSET
        if not isinstance(self.incremental_time_window, Unset):
            incremental_time_window = self.incremental_time_window.to_dict()

        can_edit = self.can_edit

        field_dict: Dict[str, Any] = {}
        field_dict.update(self.additional_properties)
        field_dict.update({})
        if connection_name is not UNSET:
            field_dict["connection_name"] = connection_name
        if target is not UNSET:
            field_dict["target"] = target
        if timestamp_columns is not UNSET:
            field_dict["timestamp_columns"] = timestamp_columns
        if incremental_time_window is not UNSET:
            field_dict["incremental_time_window"] = incremental_time_window
        if can_edit is not UNSET:
            field_dict["can_edit"] = can_edit

        return field_dict

    @classmethod
    def from_dict(cls: Type[T], src_dict: Dict[str, Any]) -> T:
        from ..models.partition_incremental_time_window_spec import (
            PartitionIncrementalTimeWindowSpec,
        )
        from ..models.physical_table_name import PhysicalTableName
        from ..models.timestamp_columns_spec import TimestampColumnsSpec

        d = src_dict.copy()
        connection_name = d.pop("connection_name", UNSET)

        _target = d.pop("target", UNSET)
        target: Union[Unset, PhysicalTableName]
        if isinstance(_target, Unset):
            target = UNSET
        else:
            target = PhysicalTableName.from_dict(_target)

        _timestamp_columns = d.pop("timestamp_columns", UNSET)
        timestamp_columns: Union[Unset, TimestampColumnsSpec]
        if isinstance(_timestamp_columns, Unset):
            timestamp_columns = UNSET
        else:
            timestamp_columns = TimestampColumnsSpec.from_dict(_timestamp_columns)

        _incremental_time_window = d.pop("incremental_time_window", UNSET)
        incremental_time_window: Union[Unset, PartitionIncrementalTimeWindowSpec]
        if isinstance(_incremental_time_window, Unset):
            incremental_time_window = UNSET
        else:
            incremental_time_window = PartitionIncrementalTimeWindowSpec.from_dict(
                _incremental_time_window
            )

        can_edit = d.pop("can_edit", UNSET)

        table_partitioning_model = cls(
            connection_name=connection_name,
            target=target,
            timestamp_columns=timestamp_columns,
            incremental_time_window=incremental_time_window,
            can_edit=can_edit,
        )

        table_partitioning_model.additional_properties = d
        return table_partitioning_model

    @property
    def additional_keys(self) -> List[str]:
        return list(self.additional_properties.keys())

    def __getitem__(self, key: str) -> Any:
        return self.additional_properties[key]

    def __setitem__(self, key: str, value: Any) -> None:
        self.additional_properties[key] = value

    def __delitem__(self, key: str) -> None:
        del self.additional_properties[key]

    def __contains__(self, key: str) -> bool:
        return key in self.additional_properties
